Add support for parental controls

pwithnall opened the pull request:
This adds support for parental controls using [libmalcontent](https://gitlab.freedesktop.org/pwithnall/malcontent/) in two ways:
 * Blacklisting of installed apps so they can’t be run by certain users
 * Preventing installation of new apps according to the OARS ratings in those apps’ appstream data

The policy is stored per-user in accountsservice. See the commit messages in this branch for more details.

This is tested and ready to merge (as far as I’m concerned), but there are a couple of areas where I suspect you will have comments/want revisions:
 * It adds a new required dependency on `libmalcontent`, and brings back the `libappstream-glib` dependency for the `common/` directory. I could make the `libmalcontent` dependency optional; but thought it would be simplest to start discussion with it required. Dropping the `libappstream-glib` dependency from `common/` again is a bit harder, as it’s needed to read the OARS ratings for apps.
 * All the decision making about parental controls happens in the polkit policy. I think it has to be this way, so that there aren’t two separate decision making points about whether to allow the current user to install a given package. For `--user` installations, the policy is applied in `flatpak-dir.c` since polkit isn’t checked for `--user` installations.
 * Currently the parental controls checks can be bypassed by setting an environment variable, which is needed to keep the tests working on CI systems which don’t have a system bus or accountsservice set up and running appropriately. It would be nice if this could be fixed, but it might involve mocking up accountsservice for the tests, which is a bit involved, and could be deferred to a follow-up MR. Suggestions welcome.

pwithnall commented:
Simple instructions for demoing it:
```
$ # This sets the parental controls policy for user ‘philip’ to allow no installation of apps with anything more than ‘none’ for realistic violence, and to blacklist running Bustle:
$ malcontent-client set philip violence-realistic=none app/org.freedesktop.Bustle/x86_64/stable
App filter for user 1000 set
$ flatpak run org.freedesktop.Bustle
error: Running app/org.freedesktop.Bustle/x86_64/stable is not allowed by the policy set by your administrator

$ flatpak --user install flathub io.github.FreeDM
Looking for matches…

io.github.FreeDM permissions:
    ipc      network     fallback-x11     pulseaudio     wayland     x11
    dri



        ID                        Arch            Branch         Remote          Download
 1. [✗] io.github.FreeDM          x86_64          stable         flathub         < 19.5 MB

Error: Installing app/io.github.FreeDM/x86_64/stable is not allowed by the policy set by your administrator
error: Failed to install io.github.FreeDM: Installing app/io.github.FreeDM/x86_64/stable is not allowed by the policy set by your administrator
```

The equivalent `--system install` command can only be demoed if you install the polkit `.rules` file from the branch.

matthiasclasen commented:
As long as you allow user installation, preventing them from running things is not going to be very effective. I can just edit the appstream data containing the oars rating, or construct the right bwrap incantation to run the app without the flatpak frontend...

matthiasclasen commented:
And thinking about the systems that are supported by flatpak (from RHEL7 to Debian stable), this will have to be an optional dependency.

nanonyme commented:
Maybe it should be possible to disable user installs through global config file?

alexlarsson commented:
@nanonyme User installation needs no privileges, so a "system config" to disable it doesn't actually buy you anything. Anyone can build their own flatpak as a user and run that with a different config. If you want to disallow users from running apps from their homedir you need to do that some other way, like disabling no-exec mounts of the user homedirs.

alexlarsson commented:
eh, *enabling* non-exec mounts that is

alexlarsson commented:
>     * brings back the `libappstream-glib` dependency for the `common/` directory

This is not a minor problem actually. This library had symbols that conflicted with some other library (libappstream i think?) which caused kde discover to crash and stuff. So, that is a big no-no.

alexlarsson commented on common/flatpak-dir.c:
```diff
@@ -7744,6 +7748,146 @@ flatpak_dir_load_appstream_store (FlatpakDir    *self,
+
+  if (!flatpak_dir_load_appstream_store (self, remote_name, arch, store,
+                                         cancellable, error))
+    return NULL;
```
This is a loading and fully parsing a huge compressed xml file each time, and also the source of the libappstream-glib dependency. It is also subject to freshness of the appstream deployment and divergence between the installed version and the remote HEAD version. 

The more correct way to handle this is to extract it from the per-app appdata xml file during deploy, like we do with the other appstream fields. See for example https://github.com/flatpak/flatpak/pull/2521 for an example of how the license field was added. We should do something similar for the the content rating field.

pwithnall replied:
Right, I’ll take a look at that. Thanks for the pointer.

pwithnall replied:
Done.

alexlarsson commented on system-helper/org.freedesktop.Flatpak.rules.in:
```diff
@@ -7,6 +7,11 @@ polkit.addRule(function(action, subject) {
+               subject.active == true && subject.local == true &&
+               action.lookup("parental-controls-repo-installation-allowed") === "1" &&
+               action.lookup("parental-controls-is-appropriate") === "1") {
+            return polkit.Result.YES;
```
I'm not sure about this, you're not applying an extra checks here. Instead you are opening a new hole for non-admins to install system-wide apps which they were not allowed to before.

pwithnall replied:
Sure, but the policy toggle for that is [disabled by default](https://gitlab.freedesktop.org/pwithnall/malcontent/blob/126190a6a9d74eba1f4390e6429e4145bc9e5ab1/accounts-service/com.endlessm.ParentalControls.AppFilter.xml#L90).

mwleeds replied:
I don't understand. Why not check `subject.isInGroup("@privileged_group@")` in this set of conditions?

pwithnall replied:
This code has significantly changed in the new version of the PR.

alexlarsson commented:
I'm a bit uncertain about this entire thing. I mean, for it to make any sense, it has to enforce that it managed to talk to some system service and got back a trustworthy response, else fail. However, that encodes a whole lot of requirements for the entire distro (has to have a accountservice running, for instance). So, in order for flatpak to run on all distros this has to be optional, either at build-time or with a system config option.

Also, the fact that everything except the system-helper checks (like all run checks and the per-user installs) are trivial to bypass in many ways (e.g. build your own flatpak, set DBUS_SYSTEM_BUS_ADDRESS, ptrace flatpak, etc) makes having access lookup failures be hard fails (with all the complexity that causes per the above) somewhat weird.

I think we should either go back to the drawing board and design a system that is enforcing, or we should just bite the bullet and accept that rating checks are optional, and if we didn't manage to get hold of some rules for the user/app we just carry on.

Of course, an enforcing system is hard. It probably involves kernel work using things like selinux, signed binaries, etc, etc. I don't expect this to be easy, although @mjg59 has expressed some interest in working on things like this for flatpak.

So, unless you want to go down the long road I think we should just make this optional which will simplify the various gymnastics to get the test running, as well as making the distro requirements problems go away.

pwithnall commented:
> … So, in order for flatpak to run on all distros this has to be optional, either at build-time or with a system config option.

Sure. As I said in the PR description, I only made it mandatory to begin with because that was the simplest starting point for discussion.

I’ll rework the branch to make the libmalcontent dependency and feature compile-time optional.

> Also, the fact that everything except the system-helper checks (like all run checks and the per-user installs) are trivial to bypass in many ways (e.g. build your own flatpak, set DBUS_SYSTEM_BUS_ADDRESS, ptrace flatpak, etc) makes having access lookup failures be hard fails (with all the complexity that causes per the above) somewhat weird.

Indeed, it’s always going to be possible to bypass this unless the system is entirely locked down (LSM, secure boot, only allowing the child to use the computer under supervision, etc.).

The idea is that this needs to not accidentally fail (for example, due to the appstream being temporarily unavailable), and it needs to be too hard for a moderately technically able child to bypass. Any child who can use a terminal is probably out of scope. Any child who can use `ptrace` or build their own flatpak should probably be hired on the spot.

If a system can be enforcing, then the eventual plan is that this feature should be able to take advantage of that. But having an enforcing system should not be a requirement.

> I think we should either go back to the drawing board and design a system that is enforcing, or we should just bite the bullet and accept that rating checks are optional, and if we didn't manage to get hold of some rules for the user/app we just carry on.

See above; I’d like to avoid accidental failures of parental controls (allowing a child to install an app they shouldn’t be able to) if possible.

> Of course, an enforcing system is hard. It probably involves kernel work using things like selinux, signed binaries, etc, etc. I don't expect this to be easy, although @mjg59 has expressed some interest in working on things like this for flatpak.

See above; that’s all definitely out of scope in the near future. But the implementation here should be able to integrate with such enforcement in future once it is implemented (and if the system provides it).

> So, unless you want to go down the long road I think we should just make this optional which will simplify the various gymnastics to get the test running, as well as making the distro requirements problems go away.

:+1:

pwithnall commented:
Pushed some updates just so they don’t get lost, but I’m going to have to pick this up again later. It requires significant rework to grab the appdata after deploy, since that means we can no longer expose the parental controls hints in the polkit check for the `Deploy()` system-helper call, which was the previous policy check point for parental controls.

Checking the appdata after deploy also means that we now have to download the entire app before we can work out whether to tell the user to get lost because they’re too young.

mwleeds commented on system-helper/flatpak-system-helper.c:
```diff
@@ -1962,15 +2065,39 @@ flatpak_authorize_method_handler (GDBusInterfaceSkeleton *interface,
+          return FALSE;
+        }
+
+      if (!authorize_deploy_add_polkit_details (installation, system_bus,
```
It's strange that we're adding the details here about whether the app is appropriate for the user but in `org.freedesktop.Flatpak.rules.in` the details are only checked if the action is `app-install` not `install-bundle`. So unless I'm missing something, the user could install a bundle that is "too extreme" for them.

pwithnall replied:
This code has significantly changed in the new version of the PR. Now the checks are done at deploy time, which covers installation and upgrade of conventional apps and bundles.

pwithnall commented:
I’ve just pushed a significant rework of this branch, which is ready for initial review, although it should be noted at this point that I haven’t given it much testing yet, so it’s not ready to merge. The approach is definitely up for review though.

The major changes are:
 * Made libmalcontent optional
 * Dropped the changes which reintroduced the libappstream-glib dependency
 * Instead of checking parental controls at the start of an install action, they’re now checked at the end of deploy, just before the deploy directory is moved to its final location. This means we have access to the `FlatpakDeploy` data which now contains the content ratings of the app, extracted from the app’s AppData by the existing machinery to parse it in flatpak. (Thanks Alex for the suggestion.)
 * This means the parental controls are checked in the same place (`flatpak_dir_deploy()`) for installs, upgrades, apps, app bundles, the user repository and the system repository/helper.
 * The authorisation decision is now made in a new polkit action, `org.freedesktop.Flatpak.parental-controls`. This is needed to allow admin overrides to allow non-privileged users to install extreme apps if it’s deemed appropriate (typically the admin would be a parent).
 * As before, none of this is real security. It’s a series of fairly large roadblocks which could be worked around by a competent user but not by an inquisitive child.

Thanks to @uajain and @andrunko for contributions to this version of the PR.

alexlarsson commented:
Can you also add changes to the CI setup to build and use libmalcontent during CI.

alexlarsson commented on common/flatpak-run.c:
```diff
@@ -3165,6 +3169,88 @@ regenerate_ld_cache (GPtrArray    *base_argv_array,
+                                            "share",
+                                            "applications",
+                                            desktop_file_name,
+                                            NULL);
```
I don't think this is necessarily a huge problem, but just a note here. An app can have multiple desktop files, but this is only using the main one. This means all desktop files in an app will have the same permissions, and that all info is taken from the main one only.

pwithnall replied:
Ah, good point. I think that’s OK though? Primarily, we expect flatpaks to be blacklisted based on their app ID, which should be present in all the `.desktop` files in the flatpak, right?

To be a bit safer, I’ll tweak it so that it always does the app ref check, rather than only doing it if it can’t do the app info check.

`mct_app_filter_is_appinfo_allowed()` is basically a convenience wrapper around several checks, which happen to all be possible with the information provided by the `GAppInfo`: https://gitlab.freedesktop.org/pwithnall/malcontent/blob/master/libmalcontent/app-filter.c#L273

Looking at the code which actually chooses which binary to run, it looks like that’s decoupled from what `.desktop` files are actually in the flatpak. Perhaps we need to add a metadata key in `flatpak-builtins-build-finish.c` which stores which `.desktop` file is the ‘main’ one associated with the command stored in `FLATPAK_METADATA_KEY_COMMAND`?

alexlarsson replied:
Actually there doesn't necessarily have to be a match of what "flatpak run" and the default desktop file runs. In practice they are probably the same, but they are really independent.

pwithnall replied:
So perhaps we should set some metadata saying what the ‘main’ desktop file is, if and only if that corresponds to the `FLATPAK_METADATA_KEY_COMMAND`. If no such metadata is set, `flatpak run` would not check any desktop file.

alexlarsson commented on common/flatpak-run.c:
```diff
@@ -3165,6 +3169,88 @@ regenerate_ld_cache (GPtrArray    *base_argv_array,
+
+  system_bus = g_bus_get_sync (G_BUS_TYPE_SYSTEM, NULL, error);
+  if (system_bus == NULL)
+    return FALSE;
```
I wonder if there is some easy check we could do that we actually got the real system bus here. Maybe we can verify that the peer uid is 0? That kills an easy route around this.

pwithnall replied:
Interesting idea. It might be possible to do that by extracting the `GSocket` from the `GSocketConnection` returned by `g_dbus_connection_get_stream()`, then calling `g_socket_get_credentials()` on it. That *should* give the credentials of the system D-Bus daemon process.

However, it’s quite a faff and would make testing significantly harder unless the code which is tested differs from that in production. I’ll see if I can fit something sensible in.

pwithnall replied:
Filed https://github.com/flatpak/flatpak/issues/2994 as a follow-up for this.

alexlarsson commented:
> Can you also add changes to the CI setup to build and use libmalcontent during CI.

I guess this is somewhat problematic wrt running the tests, but we could have a configure option to allow parental-control to be disabled by an env-var which is then only applied in the CI build.

alexlarsson commented on common/flatpak-appdata.c:
```diff
@@ -187,6 +192,70 @@ start_element (GMarkupParseContext *context,
+      /* https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html#tag-content_rating */
+      Component *component = NULL;
+      const gchar *id = NULL;
+
```
I don't like how this ignores the hierarchy of the file. We should set some in_content_rating boolean and check that in the child elements, otherwise you can create a file with e.g. a content_attribute outside a content_rating which will crash this code due to the hashtable not being allocated.

pwithnall replied:
Fixed locally.

alexlarsson commented on common/flatpak-appdata.c:
```diff
@@ -187,6 +192,70 @@ start_element (GMarkupParseContext *context,
+                                                 G_MARKUP_COLLECT_INVALID))
+            {
+              component->content_rating_type = g_strdup (type);
+              component->content_rating = g_hash_table_new_full (g_str_hash, g_str_equal, NULL, NULL);
```
Why _full when both destroynotifies are NULL?

pwithnall replied:
It’s a convention I use to make it clear that I have thought about the lifecycle of the key/value and that they are unowned. If I used `g_hash_table_new()`, it wouldn’t be as clear that I’d actually thought about it, rather than forgetting about it.

alexlarsson replied:
Ok, that is fine to me, just looked a bit weird.

alexlarsson commented on common/flatpak-appdata.c:
```diff
@@ -291,6 +368,8 @@ flatpak_parse_appdata (const char  *appdata_xml,
           *comments = g_hash_table_ref (component->comments);
           *version = g_steal_pointer (&component->version);
           *license = g_steal_pointer (&component->license);
+          *content_rating_type = g_steal_pointer (&component->content_rating_type);
```
Some user below passes null for content_rating_type and content_rating, so you need to check it here.

pwithnall replied:
Fixed locally, although I explicitly missed out the check because it’s missing from all the other arguments in that function.

alexlarsson replied:
Well, its only *really* needed in the initial commits, later commits i believe change all callers to non-NULL...

pwithnall replied:
> later commits i believe change all callers to non-NULL...

That is correct.

I’ll leave the `NULL` checks in. It’s a less surprising calling convention in any case.

alexlarsson commented on common/flatpak-dir.c:
```diff
@@ -2388,6 +2388,35 @@ flatpak_deploy_data_get_appdata_license (GVariant *deploy_data)
+{
+  g_autoptr(GVariant) metadata = g_variant_get_child_value (deploy_data, 4);
+  const char *id = NULL, *val = NULL;
+  g_autoptr(GHashTable) content_rating = g_hash_table_new_full (g_str_hash, g_str_equal, NULL, NULL);
```
No need for _full here either

pwithnall replied:
Same as above.

alexlarsson commented on common/flatpak-dir.c:
```diff
@@ -2388,6 +2388,35 @@ flatpak_deploy_data_get_appdata_license (GVariant *deploy_data)
+  g_autoptr(GHashTable) content_rating = g_hash_table_new_full (g_str_hash, g_str_equal, NULL, NULL);
+  g_autoptr(GVariantIter) iter = NULL;
+
+  g_variant_lookup (metadata, "appdata-content-rating", "(*a{ss})", NULL, &iter);
```
I think you either need to handle this returning FALSE, or the iter being NULL below, in case the key is not in the metadata.

pwithnall replied:
Good catch. Fixed locally.

alexlarsson commented on common/flatpak-dir.c:
```diff
@@ -7548,6 +7567,112 @@ apply_extra_data (FlatpakDir   *self,
+           * https://phabricator.endlessm.com/T25340 */
+          g_debug ("Skipping parental controls check for %s since the "
+                   "system bus is unavailable in the test environment", ref);
+          return TRUE;
```
a) Yes, we want a mock thing for the tests
and
b) This check is very easy to disable, maybe we can make respecting on_session a configure option that only developers enable.

pwithnall replied:
> a) Yes, we want a mock thing for the tests

Can that be deferred to some follow-up work? I will start writing it this afternoon (it needs doing for various reasons), but it’s not going to be ready before next week.

> b) This check is very easy to disable, maybe we can make respecting on_session a configure option that only developers enable.

I can do that, but again I’m a bit hesitant to be testing code which is different from what goes out into production (even if the differences are small). I think a strategy for testing flatpak (as a whole, not just the parental controls stuff) by redirecting its system bus connection needs to be looked at, and that’s not something I’ve spent any real time on yet.

Could this also be deferred to follow up work? I don’t want to rush into clamping down the system bus connection only to end up making it a complete pain to test flatpak during development.

alexlarsson replied:
Yeah, delaying both these is fine to me.

pwithnall replied:
Filed
 * https://github.com/flatpak/flatpak/issues/2993
 * https://github.com/flatpak/flatpak/issues/2994

alexlarsson commented on common/flatpak-dir.c:
```diff
@@ -7548,6 +7567,112 @@ apply_extra_data (FlatpakDir   *self,
+          return FALSE;
+        }
+    }
+
```
Also, same comment as before, maybe we can verify that the system bus is running as uid 0.

pwithnall replied:
Filed https://github.com/flatpak/flatpak/issues/2994 as a follow-up for this.

alexlarsson commented on common/flatpak-dir.c:
```diff
@@ -7548,6 +7567,112 @@ apply_extra_data (FlatpakDir   *self,
+                                           cancellable, error);
+  if (app_filter == NULL)
+    return FALSE;
+
```
I feel that we should maybe handle a failure here differently. If we are linking to libmalcontent, and we have a system bus connection, which we (somewhat) validated as trustworthy, and it turns out that there is no malcontent manager running on the system then I think this should be equivalent with "everything is allowed".

This means is should be possible to ship a flatpak linked to libmalcontent, but then chose not to install the manager and still have everything work the same as if libmalcontent was not linked to.

Does this make sense?

pwithnall replied:
Ah, this is a good catch. I actually added support for that in libmalcontent 0.3.0, but forgot to update the flatpak branch to use it. malcontent can return `MCT_APP_FILTER_ERROR_DISABLED` to indicate that filtering is disabled system-wide because, for example, the manager isn’t installed.

Fixed locally.

pwithnall commented:
I’ve pushed a load of fixups with fixes for your review comments above. One of them updates the CI to use malcontent, but I fully expect it to fail due to missing a dependency or something.

I need to test this before it’s ready to merge, and the fixups will need to be squashed. But the code changes in the fixups should be good to review.

andrunko commented on common/flatpak-run.c:
```diff
@@ -3165,6 +3169,91 @@ regenerate_ld_cache (GPtrArray    *base_argv_array,
+   * extracted from a .desktop file. */
+  allowed = mct_app_filter_is_flatpak_ref_allowed (app_filter, app_ref);
+  if (app_info != NULL)
+    allowed = allowed && mct_app_filter_is_appinfo_allowed (app_filter,
```
You may be able to save some CPU cycles here if you skip the app_info check (and even building the app_info above) if we already know the ref is not allowed.

pwithnall replied:
It makes the code a bit messier, but probably worth it to avoid reading in the desktop file.

matthiasclasen replied:
Why is there separate filtering on the app id and the appinfo in the first place?

pwithnall commented:
I’ve pushed some more fixups which rework how the polkit check happens, to avoid sending details to polkit from a non-root user, which it rejects.

The polkit check has been renamed from `parental-controls` to `override-parental-controls`, and is now only checked if the parental controls policy shows that the app shouldn’t be allowed to be installed by default. It’s basically an escape hatch to allow the administrator to override and let an app be installed (perhaps because they’ve vetted it and decided it’s OK for their child).

I’ve also re-tested it all, and pushed a few fixes as a result of that. Most notably, the `default_rating_value` in `flatpak_oars_check_rating()` is now `NONE` if an app provides ratings for other OARS sections, on the assumption that whoever wrote the appdata for that app would have included the relevant OARS sections and left out anything which had a value of `none`.

alexlarsson commented:
I'd like to take a look at this and get it merged. Can you rebase, merge the fixups and fix the conflicts?

pwithnall commented:
Rebased, squashed and conflicts fixed.

pwithnall commented:
libmalcontent requires GLib 2.60.1, which isn’t available on the CI machine (it only has 2.58.3). The commit which introduced the 2.60.1 dependency in libmalcontent is here: https://gitlab.freedesktop.org/pwithnall/malcontent/commit/d9acee829aa28256ef8c1f494b580ca5ee9585e9. I could potentially revert that (and https://gitlab.freedesktop.org/pwithnall/malcontent/commit/418b79ab). Is there a way to cleanly get GLib 2.60.1 onto the CI machine? How badly do you want libmalcontent in the CI? (We will, in any case, be building it on CI within Endless.)

mwleeds commented:
> libmalcontent requires GLib 2.60.1, which isn’t available on the CI machine (it only has 2.58.3). The commit which introduced the 2.60.1 dependency in libmalcontent is here: https://gitlab.freedesktop.org/pwithnall/malcontent/commit/d9acee829aa28256ef8c1f494b580ca5ee9585e9. I could potentially revert that (and https://gitlab.freedesktop.org/pwithnall/malcontent/commit/418b79ab). Is there a way to cleanly get GLib 2.60.1 onto the CI machine? How badly do you want libmalcontent in the CI? (We will, in any case, be building it on CI within Endless.)

This PR should get us GLib 2.60.7: https://github.com/flatpak/flatpak/pull/3088

alexlarsson commented on common/flatpak-dir.c:
```diff
@@ -7545,6 +7564,124 @@ apply_extra_data (FlatpakDir   *self,
+    }
+  else if (local_error != NULL)
+    {
+      g_propagate_error (error, g_steal_pointer (&local_error));
```
Maybe we should wrap the error in a FLATPAK_ERROR instead, so that people can match against a  new-allowed error with libflatpak without having to link to libmalcontent.

alexlarsson replied:
Or maybe these are errors only for internal errors?

pwithnall replied:
The other possible errors in this case are:
```
 * @MCT_APP_FILTER_ERROR_INVALID_USER: Given user ID doesn’t exist
 * @MCT_APP_FILTER_ERROR_PERMISSION_DENIED: Not authorized to query the app
 *    filter for the given user
 * @MCT_APP_FILTER_ERROR_INVALID_DATA: The data stored in the app filter for
 *    a user is inconsistent or invalid
```

They should probably all be transformed into a “not allowed by admin” flatpak error. I’ll do that.

alexlarsson commented on common/flatpak-dir.c:
```diff
@@ -7545,6 +7564,124 @@ apply_extra_data (FlatpakDir   *self,
+  authorized = polkit_authorization_result_get_is_authorized (result);
+
+  if (!authorized)
+    return flatpak_fail (error,
```
Anyway, this should definately add a custom flatpak error type for "not allowed by admin".

pwithnall replied:
Done.

alexlarsson commented on ci/build.sh:
```diff
@@ -9,10 +9,18 @@ dn=$(dirname $0)
 pkg_install_builddeps flatpak
 
+# malcontent isn’t packaged for our CI distributions yet
+git clone https://gitlab.freedesktop.org/pwithnall/malcontent.git /tmp/malcontent
```
We should probably specify a particular tag here rather than "whatever happens to be latest when testing the PR".

pwithnall replied:
Good suggestion, specified `0.4.0`.

pwithnall commented:
Pushed some fixup commits to address the latest review comments, plus some new commits to add the new error code. They’ll need to be reordered and the fixups squashed before this is merged.

pwithnall commented:
bot, retest this please

mwleeds commented on common/flatpak-dir.c:
```diff
@@ -7488,6 +7569,122 @@ apply_extra_data (FlatpakDir   *self,
+    {
+      /* FIXME: Instead of skipping the parental controls check in the test
+       * environment, make a mock service for it.
+       * https://phabricator.endlessm.com/T25340 */
```
I guess we need to create an upstream bug to avoid linking to phabricator.endlessm.com

pwithnall replied:
Oops, it’s https://github.com/flatpak/flatpak/issues/2993. I’ll update the link, thanks.

pwithnall commented:
Ready for review again. In the latest updates, I’ve added some documentation to the `.policy.in` file explaining how various different security policies can be enforced. (On Endless, we want a policy where any user can install ‘safe’ software, and any user who tries to install ‘unsafe’ software gets an admin prompt, even if they are in the `wheel` group. But you probably don’t want that as the default upstream unless you are going to require parental controls to be enabled.)

The other change is to fix our handling of partially-defined `<content_rating/>` elements in AppData, since the [OARS semantics have been clarified](https://github.com/hughsie/oars/pull/32).

Ready for review.

matthiasclasen commented on common/flatpak-parental-controls.c:
```diff
@@ -0,0 +1,141 @@
+    }
+
+  return TRUE;
+}
```
This loooks like it could live just as well in libmalcontent

pwithnall replied:
I think it’s a bit clearer to open-code it, since the handling of `UNKNOWN`s could vary between users of libmalcontent, and that’s basically all that’s complex here.

matthiasclasen commented on system-helper/org.freedesktop.Flatpak.policy.in:
```diff
@@ -235,5 +235,62 @@
+            once the app’s data (including its content rating) has been
+            downloaded.
+          - This action is checked to see if the installation should be allowed
+            based on whether the app being installed has extreme content.
```
Maybe 'extreme' is not the best word here, since thats just one of the possible values in OARS. 

How about 'content that is not in compliance with the users parental controls profile'. If profile is the right word. What do you call the thing you store in accountsservice ?

alexlarsson replied:
I agree that "extreme" is a poor word here, but lets land this and fix that in a followup.

pwithnall replied:
Attempted to be fixed in #3150.

alexlarsson commented on common/flatpak-dir-private.h:
```diff
@@ -32,7 +32,7 @@
  *
  * Version 1 added appdata-name/summary/version/license
- * Version 2 added extension-of
+ * Version 2 added extension-of/appdata-content-rating
```
We did release ver 2 in 1.5.0, but that is an unstable release, so i'm fine with reusing the same version here.

pwithnall replied:
You hadn’t released 1.5.0 at the time I wrote that!

alexlarsson commented on ci/build.sh:
```diff
@@ -9,10 +9,19 @@ dn=$(dirname $0)
     libubsan libasan libtsan clang python2 \
     elfutils git gettext-devel libappstream-glib-devel hicolor-icon-theme \
-    dconf-devel fuse-devel \
+    dconf-devel fuse-devel meson dbus-devel \
```
We're not building CI this way anymore. Now it happens in .github/workflows/check.yml, but lets take that in a followup.

pwithnall replied:
Attempted to be fixed in #3150.

alexlarsson commented on common/flatpak-error.h:
```diff
@@ -58,6 +58,9 @@ G_BEGIN_DECLS
  * @FLATPAK_ERROR_REF_NOT_FOUND: The specified ref was not found. (Since: 1.4.0)
+ * @FLATPAK_ERROR_PERMISSION_DENIED: An operation was not allowed by the administrative policy.
+ *                                   For example, an app is not allowed to be installed due
+ *                                   to not complying with the parental controls policy. (Since: 1.5.0)
```
Should be 1.5.1.

pwithnall replied:
Fixed in #3150.

alexlarsson commented:
@pwithnall Its time to land this now, and it seems overall fine to me. Can you do a follow-up with the 1.5.1, and docs changes, and maybe another one updating the new CI to build this.
